gateway/api/views/files.py

```diff
@@ -67,38 +69,196 @@ def check_user_has_provider(self, user, provider_name):
+        has_access = any(group in instances for group in user_groups)
+        if not has_access:
+            logger.error(
+                "User [%s] has no access to function [%s/%s].",
+                user.id,
+                function.provider.name,
+                function.title,
+            )
+        return has_access
+
     def list(self, request):
-        """List of available for user files."""
-        response = Response(
-            {"message": "Requested file was not found."},
-            status=status.HTTP_404_NOT_FOUND,
-        )
-        files = []
+        """
+        It returns a list with the names of available files for the user directory:
+            it will look under its username or username/provider_name/function_title
+        """
+
         tracer = trace.get_tracer("gateway.tracer")
         ctx = TraceContextTextMapPropagator().extract(carrier=request.headers)
         with tracer.start_as_current_span("gateway.files.list", context=ctx):
-            user_dir = request.user.username
-            provider_name = request.query_params.get("provider")
-            if provider_name is not None:
-                if self.check_user_has_provider(request.user, provider_name):
-                    user_dir = provider_name
+            username = request.user.username
+            provider_name = sanitize_name(request.query_params.get("provider"))
+            function_title = sanitize_name(request.query_params.get("function"))
+            working_dir = USER_STORAGE
+
+            function = None
+            if provider_name:
+                function = self.get_provider_function(
+                    provider_name=provider_name, function_title=function_title
+                )
+                if function:
+                    if not self.user_has_access_to_provider_function(
+                        request.user, function
+                    ):
+                        return Response(
+                            {
+                                "message": "You don't have access to this Qiskit Function."
+                            },
+                            status=status.HTTP_403_FORBIDDEN,
+                        )
                 else:
-                    return response
-            user_dir = os.path.join(
-                sanitize_file_path(settings.MEDIA_ROOT),
-                sanitize_file_path(user_dir),
-            )
-            if os.path.exists(user_dir):
-                files = [
-                    os.path.basename(path)
-                    for path in glob.glob(f"{user_dir}/*.tar")
-                    + glob.glob(f"{user_dir}/*.h5")
-                ]
+                    return Response(
+                        {
+                            "message": f"Qiskit Function {provider_name}/{function_title} doesn't exist."  # pylint: disable=line-too-long
+                        },
+                        status=status.HTTP_404_NOT_FOUND,
+                    )
```

Comment:
I see this code in a lot of lines, should we create a function?

Something like:
```python
def function_not_found_response(function_name: str, provider_name: str | None) 
    functionStr = f"{provider_name}/{function_title}" if provider_name else function_name
    return Response(
        {
            "message": f"Qiskit Function {functionStr} doesn't exist."
        },
        status=status.HTTP_404_NOT_FOUND,
    )
```

Reply:
From inner conversations we solved most of the complexity of this end-point thanks to this refactorization:
[a0210c1](https://github.com/Qiskit/qiskit-serverless/pull/1529/commits/a0210c11559762adcb917869fb9fc3dfb5d8b47a)

Another refactor to improve the current `Response` management can be done in another PR.